flow/probes/ebpf.go

```diff
@@ -394,50 +109,28 @@ func (p *EBPFProbe) run() {
 				}
 			}
 
-			kernFlow := C.struct_flow{}
 			var key, nextKey C.__u64
 			for {
-				found, err := p.module.LookupNextElement(p.fmap, unsafe.Pointer(&key), unsafe.Pointer(&nextKey), unsafe.Pointer(&kernFlow))
+				found, err := p.module.LookupNextElement(p.fmap, unsafe.Pointer(&key), unsafe.Pointer(&nextKey), unsafe.Pointer(&kernFlows[i]))
 				if !found || err != nil {
 					break
 				}
 				key = nextKey
 				// delete every entry after we read the entry value
 				p.module.DeleteElement(p.fmap, unsafe.Pointer(&key))
 
-				lastK := int64(kernFlow.last)
+				lastK := int64(kernFlows[i].last)
 				last := start.Add(time.Duration(lastK - startKTimeNs))
-
-				ebpfFlow, ok := ebpfFlows[kernFlow.key]
-				if !ok {
-					startK := int64(kernFlow.start)
-
-					// check that the local time computed from the kernel time is not greater than Now
-					us := start.Add(time.Duration(startK - startKTimeNs))
-					if us.After(now) {
-						us = now
-					}
-
-					ebpfFlow = &EBPFFlow{
-						start: now,
-						last:  last,
-					}
-					ebpfFlows[kernFlow.key] = ebpfFlow
-
-					ebpfFlow.lastK = lastK
-					ebpfFlow.last = last
-
-					for _, fops := range p.newFlowOperation(ebpfFlow, &kernFlow, startKTimeNs, start) {
-						flowChanOperation <- fops
-					}
-				} else {
-					flowChanOperation <- p.updateFlowOperation(ebpfFlow, &kernFlow, startKTimeNs, start)
+				startK := int64(kernFlows[i].start)
+				us := start.Add(time.Duration(startK - startKTimeNs))
+				if us.After(now) {
+					us = now
 				}
-			}
-
-			for k, v := range ebpfFlows {
-				if time.Now().Sub(v.last).Seconds() > p.expire.Seconds() {
-					delete(ebpfFlows, k)
+				flow.SetEBPFFlow(&ebpfFlows[i], us, last, unsafe.Pointer(&kernFlows[i]), startKTimeNs, p.probeNodeTID)
+				flowChanOperation <- &ebpfFlows[i]
+				nextAvailablePtr++
+				if nextAvailablePtr == 2001 {
```

Comment:
please set a constant like FLOW_POOL_SIZE
nextAvailablePtr is not used, should be used line 400
nextAvailablePtr = nextAvailablePtr++ % FLOW_POOL_SIZE

Reply:
you should test you got an potential overlap ( (nextAvailablePtr+1) % FLOW_POOL_SIZE == i)
meaning your pool is full

Reply:
done for the first issue.
As for the second, there shouldn't be any overlaps since the channel is FIFO and blocks when it's full (after 1000) the 2001 is merely for safety. Am I correct?